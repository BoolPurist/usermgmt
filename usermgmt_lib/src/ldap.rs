//! TODO: Implement LDAP credential Struct for centralizing username and password acquisition.

mod ldap_config;
mod ldap_credential;
mod ldap_paths;
mod ldap_search_result;
mod ldap_simple_credential;
pub mod text_list_output;

pub use ldap_config::LDAPConfig;
pub use ldap_credential::LdapCredential;
pub use ldap_search_result::LdapSearchResult;
pub use ldap_simple_credential::LdapSimpleCredential;

#[cfg(test)]
pub mod testing;
use crate::prelude::AppResult;
use crate::util::{get_new_uid, hashset_from_vec_str};
use crate::{prelude::*, NewEntity};
use crate::{ChangesToUser, MgmtConfig};
use ldap3::controls::{MakeCritical, RelaxRules};
use ldap3::{LdapConn, LdapError, LdapResult, Mod, Scope, SearchEntry};
use log::{debug, info, warn};
use maplit::hashset;
use std::collections::HashSet;

pub fn make_ldap_connection<T>(ldap_config: &LDAPConfig<T>) -> AppResult<LdapConn>
where
    T: LdapCredential,
{
    let mut ldap = LdapConn::new(&ldap_config.ldap_server)?;
    let _ = ldap
        .simple_bind(ldap_config.bind(), ldap_config.password()?)
        .with_context(|| {
            format!(
                "Failed to establish ldap connection via the bind {}",
                ldap_config.bind()
            )
        })?;
    Ok(ldap)
}

pub fn add_ldap_user<T>(
    entity: &NewEntity,
    config: &MgmtConfig,
    ldap_config: &LDAPConfig<T>,
) -> AppResult
where
    T: LdapCredential,
{
    let exitence_of_username = username_exists(entity.username.as_ref(), ldap_config)?;
    if exitence_of_username {
        warn!(
            "User {} already exists in LDAP. Skipping LDAP user creation.",
            &entity.username
        );
        return Ok(());
    }

    let uid_number = find_next_available_uid(ldap_config, entity.group.id())
        .context("No users found or LDAP query failed. Unable to assign uid. Aborting...")?;

    let mut ldap_connection = make_ldap_connection(&ldap_config)?;

    ldap_connection.simple_bind(ldap_config.bind(), ldap_config.password()?)?;
    debug!("LDAP connection established to {}", ldap_config.bind());

    add_to_ldap_db(entity, uid_number, ldap_connection, config, ldap_config)?;

    info!("Added LDAP user {}", entity.username);
    return Ok(());

    fn add_to_ldap_db<T>(
        entity: &NewEntity,
        uid: u32,
        mut ldap_connection: LdapConn,
        config: &MgmtConfig,
        ldap_config: &LDAPConfig<T>,
    ) -> AppResult
    where
        T: LdapCredential,
    {
        let un = entity.username.as_ref().as_str();
        let gid = entity.group.gid().to_string();
        let uid = uid.to_string();
        let ln = entity.lastname.as_ref().as_str();
        let gn = entity.firstname.as_ref().as_str();
        let mail: &str = entity
            .mail
            .as_ref()
            .map(|trimmmed| trimmmed.as_ref().as_str())
            .unwrap_or("");

        let def_qos = entity.default_qos.as_ref().as_str();
        let home = &format!("/home/{}", entity.username);
        let qos: HashSet<&str> = (&entity.qos)
            .into_iter()
            .map(|qos| qos.as_ref().as_str())
            .collect();
        let pubkey = entity
            .publickey
            .as_ref()
            .map(|trimmmed| trimmmed.as_ref().as_str())
            .unwrap_or("");

        let result_form_adding = ldap_connection.add(
            &format!("uid={},{}", entity.username, ldap_config.base()),
            vec![
                ("cn", hashset! {un}),
                (
                    "objectClass",
                    hashset_from_vec_str(&config.objectclass_common).to_owned(),
                ),
                ("gidNumber", hashset! {gid.as_str()}),
                ("uidNumber", hashset! {uid.as_str()}),
                ("uid", hashset! {un}),
                ("sn", hashset! {ln}),
                ("givenName", hashset! {gn}),
                ("mail", hashset! {mail}),
                ("slurmDefaultQos", hashset! {def_qos}),
                ("homeDirectory", hashset! {home.as_str()}),
                ("slurmQos", qos),
                ("sshPublicKey", hashset! {pubkey}),
                ("loginShell", hashset! {config.login_shell.as_str()}),
            ],
        );

        ldap_is_success(result_form_adding).context("Unable to create LDAP user!")?;
        Ok(())
    }
}

pub fn delete_ldap_user<T>(username: &str, ldap_config: LDAPConfig<T>) -> AppResult
where
    T: LdapCredential,
{
    // get dn for uid
    let mut ldap = make_ldap_connection(&ldap_config)?;
    let dn = find_dn_by_uid(username, &mut ldap, &ldap_config)
        .with_context(|| format!("No DN found for username {}!", username))?;
    debug!("LDAP connection established to {}", ldap_config.bind());

    match &dn {
        Some(dn_to_delete) => {
            ldap_is_success(ldap.delete(dn_to_delete)).context("User deletion in LDAP failed!")?;
            info!("Successfully deleted DN {}", dn_to_delete);
        }
        None => {
            warn!("No dn found to delete under the username {}", username);
        }
    }

    Ok(())
}

pub fn modify_ldap_user<T>(
    modifiable: &ChangesToUser,
    config: &MgmtConfig,
    ldap_config: LDAPConfig<T>,
) -> AppResult
where
    T: LdapCredential,
{
    let mut ldap = make_ldap_connection(&ldap_config)?;
    let dn = find_dn_by_uid(modifiable.username.as_ref(), &mut ldap, &ldap_config)
        .with_context(|| {
            format!(
                "No DN found for username {}! Unable to modify user.",
                modifiable.username.as_ref()
            )
        })?
        .ok_or(anyhow!("No dn found for uid"))?;

    let password = &ldap_config.password()?;
    ldap.simple_bind(ldap_config.bind(), password)?;
    debug!("LDAP connection established to {}", ldap_config.bind());

    // Prepare replace operation

    let old_qos = match &modifiable.qos {
        Some(_) => find_qos_by_uid(
            modifiable.username.as_ref(),
            config,
            ldap_config.username(),
            password,
        ),
        None => Ok(Vec::default()),
    }?;
    let mod_vec = make_modification_vec(modifiable, &old_qos);

    // Replace userPassword at given dn
    let res = ldap
        .with_controls(RelaxRules.critical())
        .modify(&dn, mod_vec);

    ldap_is_success(res).context("User modification in LDAP failed!")?;
    info!("Successfully modified user {} in LDAP", modifiable.username);
    Ok(())
}

/// List all LDAP users and some attributes
///
/// It currently outputs all values in line separated by commas.
pub fn list_ldap_users<T>(ldap_config: LDAPConfig<T>) -> AppResult<LdapSearchResult>
where
    T: LdapCredential,
{
    // Establish LDAP connection and bind
    let mut ldap =
        make_ldap_connection(&ldap_config).context("Error while connecting via LDAP !")?;

    debug!(
        "LDAP connection established to {}. Will search under {}",
        ldap_config.bind(),
        ldap_config.base()
    );
    let attrs = {
        // Make sure the keys are sorted alphabetic
        // This way the order fields in the final output deterministic
        let mut to_sort = vec![
            "uid",
            "uidNumber",
            "givenName",
            "sn",
            "mail",
            "slurmDefaultQos",
            "slurmQos",
        ];
        to_sort.sort();
        to_sort
    };

    // Search for all entities under base dn
    let search_result = ldap
        .search(
            ldap_config.base(),
            Scope::OneLevel,
            "(objectclass=*)",
            attrs.clone(),
        )
        .context("Error during LDAP search!")?;

    let search_result = LdapSearchResult::from_ldap_raw_search(&attrs, &search_result);

    Ok(search_result)
}

fn make_modification_vec<'a>(
    modifiable: &'a ChangesToUser,
    old_qos: &'a Vec<String>,
) -> Vec<Mod<&'a str>> {
    macro_rules! may_push_simple_modification {
        ($name:expr, $modifable:ident, $modification:ident, $field:ident) => {
            if let Some(val) = &$modifable.$field {
                info_log($name);
                ($modification).push(Mod::Replace($name, HashSet::from([val.as_ref().as_str()])))
            }
        };
    }
    let mut modifications: Vec<Mod<&str>> = Vec::new();

    let modifiable = modifiable.as_ref();
    may_push_simple_modification!("givenName", modifiable, modifications, firstname);
    may_push_simple_modification!("sn", modifiable, modifications, lastname);
    may_push_simple_modification!("mail", modifiable, modifications, mail);
    may_push_simple_modification!("slurmDefaultQos", modifiable, modifications, default_qos);
    may_push_simple_modification!("publickey", modifiable, modifications, publickey);

    let replace_old_with_new_qos = !old_qos.is_empty();
    if replace_old_with_new_qos {
        // first we delete all old qos
        const SLURM_QOS: &str = "slurmQos";
        info_log(SLURM_QOS);
        for q in old_qos {
            modifications.push(Mod::Delete(SLURM_QOS, HashSet::from([q.as_str()])))
        }
        // then we add all new qos
        for q in modifiable.qos.iter() {
            let q: HashSet<&str> = q.into_iter().map(|qos| qos.as_ref().as_str()).collect();
            modifications.push(Mod::Add(SLURM_QOS, q))
        }
    }
    return modifications;

    fn info_log(field: &str) {
        info!("Changing the field: {}", field)
    }
}

/// Do a LDAP search to determine the next available uid
fn find_next_available_uid<T>(ldap_config: &LDAPConfig<T>, group: crate::Group) -> AppResult<u32>
where
    T: LdapCredential,
{
    let mut ldap = make_ldap_connection(&ldap_config).context("Error during uid search!")?;

    let password = ldap_config.password()?;
    debug!("Binding with dn: {}, pw: {}", ldap_config.bind(), password);

    debug!(
        "find_next_available_uid: LDAP connection established to {}",
        ldap_config.bind(),
    );

    debug!("Search under {}", ldap_config.base());

    // Search for all uidNumbers under base dn
    let search_result = ldap
        .search(
            ldap_config.base(),
            Scope::OneLevel,
            "(objectclass=*)",
            vec!["uidNumber"],
        )
        .context("Error during uid search!")?;

    let mut uids: Vec<u32> = Vec::new();
    for elem in search_result.0.iter() {
        let search_result = SearchEntry::construct(elem.to_owned());
        debug!("UID: {:?}", SearchEntry::construct(elem.to_owned()));
        let uid = &search_result.attrs["uidNumber"][0].parse::<u32>().unwrap();
        uids.push(*uid);
    }

    get_new_uid(&uids, group)
}

/// Search for a specific uid and return the corresponding dn.
fn find_dn_by_uid<T>(
    username: &str,
    ldap: &mut LdapConn,
    ldap_config: &LDAPConfig<T>,
) -> AppResult<Option<String>>
where
    T: LdapCredential,
{
    debug!("LDAP connection established to {}", ldap_config.bind());

    // Search for all uids under base dn and return dn of user
    let search = ldap.search(
        ldap_config.base(),
        Scope::OneLevel,
        &format!("(uid={username})"),
        vec!["dn"],
    )?;

    let entry = search
        .0
        .into_iter()
        .next()
        .with_context(|| format!("No LDAP entry found for user {}", username))?;

    let sr = SearchEntry::construct(entry);
    debug!("SR for deletion: {:?}", sr);

    Ok(Some(sr.dn))
}

/// Search for a specific uid and return the corresponding qos.
fn find_qos_by_uid(
    username: &str,
    config: &MgmtConfig,
    ldap_user: &str,
    ldap_pass: &str,
) -> AppResult<Vec<String>> {
    let ldap_config = LDAPConfig::new(
        config,
        LdapSimpleCredential::new(ldap_user.to_string(), ldap_pass.to_string()),
    )?;
    let mut fetched_all_qos: Vec<String> = Vec::new();

    let ldap_server = &ldap_config.ldap_server;
    let mut ldap_connection = make_ldap_connection(&ldap_config)
        .with_context(|| format!("Connection to {} failed", ldap_server))?;

    debug!("LDAP connection established to {}", ldap_config.bind());

    // Search for all uid under base dn and return dn of user
    let search = ldap_connection
        .search(
            ldap_config.base(),
            Scope::OneLevel,
            &format!("(uid={})", username),
            vec!["slurmQos"],
        )
        .with_context(|| {
            format!(
                "search did not find any slurmQos for the user with uid {}",
                username
            )
        })?;

    for elem in search.0.iter() {
        let search_result = SearchEntry::construct(elem.to_owned());
        let q = &search_result.attrs["slurmQos"];
        for one_qos in q {
            debug!("Fetched QOS: {:?}", one_qos);
            fetched_all_qos.push(one_qos.clone());
        }
    }

    Ok(fetched_all_qos)
}

/// Check if username already exists in ldap.
/// Must be an exact match on the uid attribute.
fn username_exists<T>(username: &String, ldap_config: &LDAPConfig<T>) -> AppResult<bool>
where
    T: LdapCredential,
{
    let mut username_exists = false;
    let mut ldap = make_ldap_connection(&ldap_config)?;
    debug!("LDAP connection established to {}", ldap_config.bind());

    // Search for all uid under base dn and return dn of user
    let search_result = ldap.search(
        ldap_config.base(),
        Scope::OneLevel,
        &format!("(uid={username})"),
        vec!["dn"],
    )?;
    match search_result.0.into_iter().next() {
        Some(entry) => {
            // User found. Good.
            debug!("Found user: {:?}", SearchEntry::construct(entry));
            username_exists = true
        }
        None => debug!("No LDAP entry found for user {}", username),
    }
    Ok(username_exists)
}

/// If ok is returned then ldap operation happened with zero error code, LDAP_SUCCESS
///
/// Even if a call to ldap returns ok it has an error code inside it. Only if the code is zero
/// then the operation really happened successfully.
/// Link: https://docs.rs/ldap3/latest/ldap3/result/struct.LdapResult.html
fn ldap_is_success(to_check: Result<LdapResult, LdapError>) -> Result<(), LdapError> {
    match to_check {
        Ok(might_have_non_zero_error_code) => match might_have_non_zero_error_code.success() {
            Ok(_with_zero_error_code) => Ok(()),
            Err(error) => Err(error),
        },
        Err(error) => Err(error),
    }
}

fn ask_credentials_if_not_provided<T>(
    username: Option<&str>,
    password: Option<&str>,
    credential: &T,
) -> AppResult<(String, String)>
where
    T: LdapCredential,
{
    let (ldap_user, ldap_pass) = match (username, password) {
        (None, None) => (credential.username()?, credential.password()?),
        (Some(username), None) => (username, credential.password()?),
        (None, Some(password)) => (credential.username()?, password),
        (Some(username), Some(password)) => (username, password),
    };

    return Ok((ldap_user.trim().to_owned(), ldap_pass.trim().to_owned()));
}
