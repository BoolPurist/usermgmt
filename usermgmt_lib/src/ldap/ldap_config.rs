use crate::{prelude::AppResult, MgmtConfig};

use super::{ldap_paths::LdapPaths, LdapCredential};
#[derive(Debug, Default)]
/// Contains all information for creating/deleting and modifying an user aka writting actions
/// TODO: consider implementing encapsulation with getters and setters
pub struct LDAPConfig<T> {
    pub ldap_server: String,
    ldap_credentails: T,
    pub ldap_paths: LdapPaths,
}

impl<T> LDAPConfig<T>
where
    T: LdapCredential,
{
    pub fn new(config: &MgmtConfig, credentials: T) -> AppResult<Self> {
        let (bind_prefix, ldap_server, dc, org_unit, bind_org_unit) = (
            &config.ldap_bind_prefix,
            &config.ldap_server,
            &config.ldap_domain_components,
            &config.ldap_org_unit,
            &config.ldap_bind_org_unit,
        );

        let ldap_user = credentials.username()?;

        let ldap_paths = LdapPaths::new(
            dc.clone(),
            org_unit.clone(),
            bind_org_unit.clone(),
            bind_prefix.clone(),
            ldap_user.to_string(),
        );

        Ok(Self {
            ldap_server: ldap_server.to_string(),
            ldap_credentails: credentials,
            ldap_paths,
        })
    }

    pub fn bind(&self) -> &str {
        self.ldap_paths.bind()
    }
    pub fn base(&self) -> &str {
        self.ldap_paths.base()
    }
    pub fn username(&self) -> &str {
        self.ldap_paths.username()
    }
    pub fn password(&self) -> AppResult<&str> {
        self.ldap_credentails.password()
    }
}
