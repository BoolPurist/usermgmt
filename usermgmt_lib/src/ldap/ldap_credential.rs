use crate::prelude::AppResult;

pub trait LdapCredential {
    fn username(&self) -> AppResult<&str>;
    fn password(&self) -> AppResult<&str>;
    fn set_password(&mut self, new: String);
}
