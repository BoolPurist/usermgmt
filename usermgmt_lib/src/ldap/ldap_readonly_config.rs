use crate::{config::MgmtConfig, prelude::AppResult};

use super::ldap_paths::LdapPaths;
use getset::Getters;
use log::info;

#[derive(Getters, Debug)]
#[getset(get = "pub")]
/// Used for action over connection with read only access
pub struct LdapReadonlyConfig {
    ldap_server: String,
    ldap_pass: String,
    ldap_paths: LdapPaths,
}

impl LdapReadonlyConfig {
    pub fn new(config: &MgmtConfig) -> AppResult<Self> {
        Self::create(config, super::ask_credentials_in_tty)
    }

    fn create(
        config: &MgmtConfig,
        on_credentials: impl FnOnce() -> AppResult<(String, String)>,
    ) -> AppResult<Self> {
        let ldap_server = config.ldap_server.clone();
        let (ldap_user, ldap_pass) = super::ask_credentials_if_not_provided(
            config.ldap_readonly_user.as_deref(),
            config.ldap_readonly_pw.as_deref(),
            on_credentials,
        )?;

        let (bind, prefix) = (
            config.ldap_readonly_bind.clone().or_else(|| {
                info!(
                    "No org bind for readonly user provided, falling back to normal user bind org unit."
                );
                config.ldap_bind_org_unit.clone()
            }),
            config.ldap_readonly_user_prefix.clone().or_else(|| {
                info!("No prefix for readonly user provided, falling back to normal user prefix.");
                config.ldap_bind_prefix.clone()
            }),
        );
        let ldap_paths = LdapPaths::new(
            config.ldap_domain_components.clone(),
            config.ldap_org_unit.clone(),
            bind,
            prefix,
            ldap_user,
        );

        Ok(Self {
            ldap_paths,
            ldap_pass,
            ldap_server,
        })
    }

    pub fn bind(&self) -> &str {
        self.ldap_paths.bind()
    }
    pub fn base(&self) -> &str {
        self.ldap_paths.base()
    }
}
#[cfg(test)]
mod testing {
    use super::*;
    #[test]
    fn fall_back_for_default() {
        let given = LdapReadonlyConfig::create(&MgmtConfig::default(), || {
            Ok(("user".to_owned(), "password".to_owned()))
        })
        .unwrap();
        assert_eq!("password", given.ldap_pass);
        assert_eq!("user", given.ldap_paths.username());
    }
    #[test]
    fn use_all_readonly_options() {
        const USER: &str = "readonly_user";
        const PASSWORD: &str = "readonly_password";
        const BIND: &str = "ou=people,ou=department";
        const PREFIX: &str = "zzz";
        let config = MgmtConfig {
            ldap_readonly_user: Some(USER.to_owned()),
            ldap_readonly_pw: Some(PASSWORD.to_owned()),
            ldap_readonly_bind: Some(BIND.to_owned()),
            ldap_domain_components: Some("dc=example,dc=com".to_owned()),
            ldap_readonly_user_prefix: Some(PREFIX.to_owned()),
            ..Default::default()
        };
        let given = LdapReadonlyConfig::create(&config, || {
            panic!("Should fall back for asking credentials")
        })
        .unwrap();
        assert_eq!(PASSWORD, given.ldap_pass);
        assert_eq!(USER, given.ldap_paths.username());
        let expected_bind = format!(
            "{}={},{},{}",
            PREFIX,
            USER,
            BIND,
            config.ldap_domain_components.unwrap()
        );
        assert_eq!(expected_bind, given.bind());
    }
}
