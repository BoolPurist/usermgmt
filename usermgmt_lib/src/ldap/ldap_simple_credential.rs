use crate::prelude::AppResult;

use super::LdapCredential;

#[derive(Debug)]
pub struct LdapSimpleCredential {
    username: String,
    password: String,
}

impl LdapCredential for LdapSimpleCredential {
    fn username(&self) -> AppResult<&str> {
        Ok(&self.username)
    }

    fn password(&self) -> AppResult<&str> {
        Ok(&self.password)
    }
}
