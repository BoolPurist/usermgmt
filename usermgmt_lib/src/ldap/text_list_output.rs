use super::ldap_search_result::LdapSearchResult;

/// Returns rows. Every row consists comma separated cells. Every cell is key value pair with "="
/// sign in the middle.
///
/// Example for cell: name=example
///
/// If there is no value for the field, then there is nothing from right of the "=" sign
pub fn ldap_simple_output(search_results: &LdapSearchResult) -> String {
    let attrs = search_results.fields();
    let search_by_in_title = search_results.headers();
    attrs
        .iter()
        .map(|next_row| {
            next_row
                .iter()
                .enumerate()
                .map(|(index, next_field)| {
                    let title = *search_by_in_title
                        .get(index)
                        .expect("Index comes from iterator");
                    format!("{}={}", title, next_field.join("|"))
                })
                .collect::<Vec<String>>()
                .join(",")
        })
        .collect::<Vec<String>>()
        .join("\n")
}

/// Returns a pretty ASCII table from the result of an LDAP search query.
/// First row is the title row which shows all used field name.
/// Subsequent rows contain values to each field name for an LDAP entity.
///
/// If there is no value for the field, then the cell of in the column is white space only
pub fn ldap_search_to_pretty_table(search_result: &LdapSearchResult) -> String {
    use prettytable::{Cell, Row, Table};

    let headers = search_result.headers();
    let mut table = table_with_title_bar(headers.as_slice());

    for row_to_convert in search_result.fields().iter() {
        // Construct with initial empty cells
        let mut cells = vec![Cell::new(""); headers.len()];

        for (field_index, cell_value) in row_to_convert.iter().enumerate() {
            let to_mutate = cells
                .get_mut(field_index)
                .expect("Index comes from iterator");
            *to_mutate = Cell::new(&cell_value.join(" | "));
        }

        table.add_row(Row::new(cells));
    }

    return table.to_string();

    fn table_with_title_bar(fiedl_names: &[&str]) -> Table {
        let mut table = Table::new();

        // Title bar
        let title_cells = fiedl_names
            .iter()
            .map(|to_cell| Cell::new(to_cell))
            .collect();

        table.set_titles(Row::new(title_cells));

        table
    }
}
