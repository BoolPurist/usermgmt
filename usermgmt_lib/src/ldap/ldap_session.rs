use crate::{AppError, AppResult};
use ldap3::LdapConn;

use crate::config::MgmtConfig;
use crate::ldap;

use super::{LDAPConfig, LdapCredential};

pub struct LdapSession<T> {
    config: LDAPConfig<T>,
    connection: Option<AppResult<LdapConn>>,
}

impl<T> LdapSession<T>
where
    T: LdapCredential,
{
    pub fn new(config: &MgmtConfig, credentials: T) -> AppResult<Self> {
        let config = LDAPConfig::new(config, credentials)?;
        let connection = None;
        Ok(Self { config, connection })
    }

    pub fn config(&self) -> &LDAPConfig<T> {
        &self.config
    }

    pub fn action<RT>(
        &mut self,
        action: impl FnOnce(&mut LdapConn, &LDAPConfig<T>) -> AppResult<RT>,
    ) -> Result<RT, AppError> {
        self.establish_connection()?;
        let connection = self.connection.as_mut().unwrap().as_mut().unwrap();
        action(connection, &self.config)
    }

    pub fn establish_connection(&mut self) -> Result<(), AppError> {
        match self.connection.as_mut() {
            Some(Ok(_)) => Ok(()),
            Some(Err(error)) => Err(anyhow::format_err!(
                "Establishing connection failed\n{:?}",
                error
            )),
            None => {
                let connection = ldap::make_ldap_connection(&self.config)?;
                self.connection = Some(Ok(connection));
                Ok(())
            }
        }
    }
}
