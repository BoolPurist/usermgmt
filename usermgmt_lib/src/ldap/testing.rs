use crate::ldap::{
    ldap_search_result::LdapSearchResult, ldap_simple_credential::LdapSimpleCredential,
};

use super::*;
use maplit::hashmap;

struct ExpectedLdapPaths {
    pub ldap_bind: String,
    pub ldap_base: String,
}

#[test]
fn should_give_correct_ldap_paths() {
    {
        let given = MgmtConfig {
            ..MgmtConfig::default()
        };
        let actual = LDAPConfig::new(
            &given,
            LdapSimpleCredential::new("xxx".to_owned(), "xxxx".to_owned()),
        )
        .unwrap();
        assert_case(
            ExpectedLdapPaths {
                ldap_base: "".to_owned(),
                ldap_bind: "cn=xxx".to_owned(),
            },
            &actual,
        );
    }

    {
        let given = MgmtConfig {
            ldap_domain_components: Some("dc=example,dc=com".to_owned()),
            ldap_org_unit: Some("ou=it,ou=people".to_owned()),
            ldap_bind_org_unit: Some("ou=special".to_owned()),

            ..MgmtConfig::default()
        };

        let actual = LDAPConfig::new(
            &given,
            LdapSimpleCredential::new("alice".to_owned(), "xxxx".to_owned()),
        )
        .unwrap();

        assert_case(
            ExpectedLdapPaths {
                ldap_base: "ou=it,ou=people,dc=example,dc=com".to_owned(),
                ldap_bind: "cn=alice,ou=special,dc=example,dc=com".to_owned(),
            },
            &actual,
        );
    }

    {
        let given = MgmtConfig {
            ldap_domain_components: Some("dc=example,dc=com".to_owned()),
            ..MgmtConfig::default()
        };
        let actual = LDAPConfig::new(
            &given,
            LdapSimpleCredential::new("alice".to_owned(), "xxxx".to_owned()),
        )
        .unwrap();
        assert_case(
            ExpectedLdapPaths {
                ldap_base: "dc=example,dc=com".to_owned(),
                ldap_bind: "cn=alice,dc=example,dc=com".to_owned(),
            },
            &actual,
        );
    }

    {
        let given = MgmtConfig {
            ldap_domain_components: Some("dc=example,dc=com".to_owned()),
            ldap_org_unit: Some("aaa=department,bbb=level".to_owned()),
            ldap_bind_prefix: Some("uid".to_string()),
            ..MgmtConfig::default()
        };
        let actual = LDAPConfig::new(
            &given,
            LdapSimpleCredential::new("alice".to_owned(), "xxxx".to_owned()),
        )
        .unwrap();
        assert_case(
            ExpectedLdapPaths {
                ldap_base: "aaa=department,bbb=level,dc=example,dc=com".to_owned(),
                ldap_bind: "uid=alice,dc=example,dc=com".to_owned(),
            },
            &actual,
        );
    }

    fn assert_case(expected: ExpectedLdapPaths, actual: &LDAPConfig<LdapSimpleCredential>) {
        assert_eq!(
            expected.ldap_base,
            actual.base(),
            "Did not produce correct base with  org path + dc."
        );
        assert_eq!(
            expected.ldap_bind,
            actual.bind(),
            "Correct user binding, prefix_user + user + bind_org + dc"
        );
    }
}

#[test]
fn should_produce_simple_output() {
    let given_entries = vec![
        hashmap! {
            "qos".to_string() => vec!["basic".to_string(), "default".to_string()],
            "name".to_string() => vec!["Mr. X".to_string()],
            "age".to_string() => vec!["2".to_string()]
        },
        hashmap! {
            "name".to_string() => vec!["example".to_string()]
        },
        hashmap! {
            "qos".to_string() => vec!["default".to_string()],
            "name".to_string() => vec!["example_man".to_string()],
        },
    ];
    let given_search_entries = vec!["qos", "name"];
    let ldap_search_result = LdapSearchResult::new(given_search_entries, given_entries);
    let actual = text_list_output::ldap_simple_output(&ldap_search_result);
    insta::assert_display_snapshot!(actual);
}

#[test]
fn should_produce_table_from_ldap_search() {
    let given_entries = vec![
        hashmap! {
            "qos".to_string() => vec!["basic".to_string(), "default".to_string()],
            "name".to_string() => vec!["Mr. X".to_string()],
            "age".to_string() => vec!["2".to_string()]
        },
        hashmap! {
            "name".to_string() => vec!["example".to_string()]
        },
    ];

    let given_search_entries = vec!["qos", "name"];
    let ldap_search_result = LdapSearchResult::new(given_search_entries, given_entries);
    let actual = text_list_output::ldap_search_to_pretty_table(&ldap_search_result);
    insta::assert_display_snapshot!(actual);
}
