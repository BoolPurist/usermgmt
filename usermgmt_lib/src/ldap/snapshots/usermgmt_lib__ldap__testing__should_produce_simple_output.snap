---
source: usermgmt_lib/src/ldap/testing.rs
expression: actual
---
qos=basic|default,name=Mr. X
qos=,name=example
qos=default,name=example_man
